File: param_gauss_recon/Method/kernel.py
import torch
from tqdm import trange

from param_gauss_recon.Config.constant import EPSILON
from param_gauss_recon.Data.pgr_params import PGRParams


def get_A(x: torch.Tensor, y: torch.Tensor, x_width: torch.Tensor):
    """
    x: numpy array of shape [N_query, 3]
    y: numpy array of shape [N_sample, 3]
    x_width: [N_query]
    ---
    return:
    A: numpy array of shape [N_query, 3 * N_sample]
    """

    N_query = x.shape[0]

    A = x[:, None] - y[None, :]  # [N_query, N_sample, 3]
    dist = torch.sqrt((A**2).sum(-1))  # [N_query, N_sample], |x^i-y^j|^2
    dist[dist == 0.0] += EPSILON

    # inv_dist = cnp.where(dist > x_width[:, None], 1/dist, 0.) # / 4 / cp.pi
    inv_dist = torch.where(
        dist > x_width[:, None], 1 / dist, 1 / x_width[:, None]
    )  # / 4 / cp.pi
    inv_cub_dist = inv_dist**3 / 4 / torch.pi  # [N_query, N_sample]

    A *= inv_cub_dist[..., None]  # [N_query, N_sample, 3]
    A = A.permute(0, 2, 1)
    A = A.reshape(N_query, -1)

    return -A

def mul_A_T(x: torch.Tensor, y: torch.Tensor, xi: torch.Tensor,
            x_width: torch.Tensor, chunk_size: int):
    N_sample = y.shape[0]
    n_y_chunks = N_sample // chunk_size + 1

    lse = torch.zeros((3, N_sample), dtype=x.dtype, device=x.device)
    for i in range(n_y_chunks):
        y_chunk = y[i * chunk_size : (i + 1) * chunk_size]
        A_chunk = get_A(x, y_chunk, x_width)
        (
            lse[0, i * chunk_size : (i + 1) * chunk_size],
            lse[1, i * chunk_size : (i + 1) * chunk_size],
            lse[2, i * chunk_size : (i + 1) * chunk_size],
        ) = torch.einsum("jk,j", A_chunk, xi).reshape(3, -1)

    return lse.reshape(-1)

def get_B(x: torch.Tensor, y: torch.Tensor,
          chunk_size: int, x_width: torch.Tensor, alpha: float):
    """
    x: numpy array of shape [N_query, 3]
    y: numpy array of shape [N_sample, 3]
    ---
    return:
    B: numpy array of shape [N_query, N_query], which is AA^T
    """
    N_query = x.shape[0]

    B = torch.zeros([N_query, N_query], dtype=x.dtype, device=x.device)

    n_row_chunks = N_query // chunk_size + 1
    n_col_chunks = n_row_chunks

    for i in trange(n_row_chunks):
        x_chunk_i = x[i * chunk_size : (i + 1) * chunk_size]

        if x_chunk_i.shape[0] <= 0:
            continue

        x_chunk_eps_i = x_width[i * chunk_size : (i + 1) * chunk_size]
        A_block_i = get_A(x_chunk_i, y, x_chunk_eps_i)

        for j in range(i, n_col_chunks):
            x_chunk_j = x[j * chunk_size : (j + 1) * chunk_size]

            if x_chunk_j.shape[0] <= 0:
                continue

            x_chunk_eps_j = x_width[j * chunk_size : (j + 1) * chunk_size]
            A_block_j = get_A(x_chunk_j, y, x_chunk_eps_j)

            B[
                i * chunk_size : (i + 1) * chunk_size,
                j * chunk_size : (j + 1) * chunk_size,
            ] = torch.einsum("ik,jk->ij", A_block_i, A_block_j)

            if i != j:
                B[
                    j * chunk_size : (j + 1) * chunk_size,
                    i * chunk_size : (i + 1) * chunk_size,
                ] = B[
                    i * chunk_size : (i + 1) * chunk_size,
                    j * chunk_size : (j + 1) * chunk_size,
                ].T
            else:
                block_size = B[
                    i * chunk_size : (i + 1) * chunk_size,
                    j * chunk_size : (j + 1) * chunk_size,
                ].shape[0]
                if block_size <= 0:
                    continue
                diag_mask = torch.eye(block_size, dtype=torch.bool, device=x.device)
                # diag_sqr = (B[i*chunk_size:(i+1)*chunk_size, j*chunk_size:(j+1)*chunk_size][diag_mask])
                B[
                    i * chunk_size : (i + 1) * chunk_size,
                    j * chunk_size : (j + 1) * chunk_size,
                ][diag_mask] *= alpha

    return B

def solveLSE(
    x: torch.Tensor, y: torch.Tensor, x_width: torch.Tensor,
    chunk_size: int, iso_value: float, r_sq_stop_eps: float,
    pgr_params: PGRParams):
    """
    x: numpy array of shape [N_query, 3]
    y: numpy array of shape [N_sample, 3]
    ---
    return:
    lse:
    """
    print("[INFO][kernel::solveLSE]")
    print('\t start pre-computing B...')
    B = get_B(x, y, chunk_size, x_width, pgr_params.alpha)

    xi = torch.zeros(x.shape[0], dtype=x.dtype, device=x.device)
    r = torch.ones(x.shape[0], dtype=x.dtype, device=x.device) * iso_value
    p = r.clone()

    print("[INFO][kernel::solveLSE]")
    print('\t start CG iterations...')
    solve_progress_bar = trange(y.shape[0])
    for _ in solve_progress_bar:
        Bp = torch.matmul(B, p)
        r_sq = r.dot(r)

        alpha = r_sq / p.dot(Bp)
        xi += alpha * p
        r -= alpha * Bp
        beta = r.dot(r) / r_sq
        p *= beta
        p += r

        solve_progress_bar.set_description(
            f"[In solver] {r_sq.item():.2e}/{r_sq_stop_eps:.0e}"
        )

        if r_sq < r_sq_stop_eps:
            solve_progress_bar.close()
            break
    lse = mul_A_T(x, y, xi, x_width, chunk_size)

    return lse

File: src/kernel.cpp
#include "constant.h"
#include "kernel.h"

using namespace torch::indexing;

const torch::Tensor get_A(
    const torch::Tensor &x,
    const torch::Tensor &y,
    const torch::Tensor &x_width){
  const int N_query = x.size(0);

  torch::Tensor A = x.unsqueeze(1) - y.unsqueeze(0);

  torch::Tensor dist = torch::sqrt((A * A).sum(-1));
  dist.index_put_({dist == 0.0}, EPSILON);

  const torch::Tensor x_width_unsqueezed = x_width.unsqueeze(1);
  const torch::Tensor inv_dist = torch::where(dist > x_width_unsqueezed, 1.0 / dist, 1.0 / x_width_unsqueezed);
  const torch::Tensor inv_cub_dist = inv_dist * inv_dist * inv_dist / 4.0 / M_PI;

  A *= inv_cub_dist.unsqueeze(-1);
  A = A.permute({0, 2, 1});
  A = A.reshape({N_query, -1});

  return -A;
}

const torch::Tensor mul_A_T(
    const torch::Tensor &x,
    const torch::Tensor &y,
    const torch::Tensor &xi,
    const torch::Tensor &x_width,
    const int &chunk_size){
  const int N_sample = y.size(0);
  const int n_y_chunks = std::floor(N_sample / chunk_size) + 1;

  const torch::TensorOptions opts = torch::TensorOptions().dtype(x.dtype()).device(x.device());
  torch::Tensor lse = torch::zeros({3, N_sample}, opts);
  for (int i = 0; i < n_y_chunks; ++i){
    const Slice chunk_slice = Slice(i * chunk_size, (i + 1) * chunk_size);
    const torch::Tensor y_chunk = y.index({chunk_slice});
    const torch::Tensor A_chunk = get_A(x, y_chunk, x_width);

    const torch::Tensor lse_chunk_results = torch::einsum("jk,j", {A_chunk, xi}).reshape({3, -1});
    for (int j = 0; j < 3; ++j){
      lse.index_put_({j, chunk_slice}, lse_chunk_results[j]);
    }
  }

  return lse.reshape({-1});
}

const torch::Tensor get_B(
    const torch::Tensor &x,
    const torch::Tensor &y,
    const int &chunk_size,
    const torch::Tensor &x_width,
    const float &alpha){
  const int N_query = x.size(0);

  const torch::TensorOptions opts = torch::TensorOptions().dtype(x.dtype()).device(x.device());
  const torch::TensorOptions bool_opts = torch::TensorOptions().dtype(torch::kBool).device(x.device());
  torch::Tensor B = torch::zeros({N_query, N_query}, opts);

  const int n_row_chunks = std::floor(N_query / chunk_size) + 1;
  const int n_col_chunks = n_row_chunks;

  for (int i = 0; i < n_row_chunks; ++i) {
    const Slice row_chunk_slice = Slice(i * chunk_size, (i + 1) * chunk_size);
    const torch::Tensor x_chunk_i = x.index({row_chunk_slice});

    if (x_chunk_i.size(0) <= 0){
      continue;
    }

    const torch::Tensor x_chunk_eps_i = x_width.index({row_chunk_slice});
    const torch::Tensor A_block_i = get_A(x_chunk_i, y, x_chunk_eps_i);

    for (int j = i; j < n_col_chunks; ++j){
      const Slice col_chunk_slice = Slice(j * chunk_size, (j + 1) * chunk_size);
      const torch::Tensor x_chunk_j = x.index({col_chunk_slice});

      if (x_chunk_j.size(0) <= 0){
        continue;
      }

      const torch::Tensor x_chunk_eps_j = x_width.index({col_chunk_slice});
      const torch::Tensor A_block_j = get_A(x_chunk_j, y, x_chunk_eps_j);

      B.index_put_({row_chunk_slice, col_chunk_slice}, torch::einsum("ik,jk->ij", {A_block_i, A_block_j}));

      if (i != j){
        B.index_put_({col_chunk_slice, row_chunk_slice}, B.index({row_chunk_slice, col_chunk_slice}).transpose(1, 0));
      }
      else{
        const int block_size = B.index({row_chunk_slice, col_chunk_slice}).size(0);
        if (block_size <= 0){
          continue;
        }

        const torch::Tensor diag_mask = torch::eye(block_size, bool_opts);
        torch::Tensor B_block = B.index({row_chunk_slice, col_chunk_slice});
        B_block.index_put_({diag_mask}, B_block.index({diag_mask}) * alpha);
        B.index_put_({row_chunk_slice, col_chunk_slice}, B_block);
      }
    }
  }

  return B;
}

const torch::Tensor solve(
    const torch::Tensor &x,
    const torch::Tensor &y,
    const torch::Tensor &x_width,
    const int &chunk_size,
    const float &iso_value,
    const float &r_sq_stop_eps,
    const PGRParams &pgr_params){
  std::cout << "[INFO][kernel::solve]" << std::endl;
  std::cout << "\t start pre-computing B..." << std::endl;
  const torch::Tensor B = get_B(x, y, chunk_size, x_width, pgr_params.alpha);

  const torch::TensorOptions opts = torch::TensorOptions().dtype(x.dtype()).device(x.device());
  torch::Tensor xi = torch::zeros({x.size(0)}, opts);
  torch::Tensor r = torch::ones({x.size(0)}, opts) * iso_value;
  torch::Tensor p = r.clone();

  std::cout << "[INFO][kernel::solve]" << std::endl;
  std::cout << "\t start CG iterations..." << std::endl;
  for (int i = 0; i < y.size(0); ++i){
    const torch::Tensor Bp = torch::matmul(B, p);
    const float r_sq = r.dot(r).item<float>();

    const float alpha = r_sq / p.dot(Bp).item<float>();
    xi += alpha * p;
    r -= alpha * Bp;
    const float beta = r.dot(r).item<float>() / r_sq;
    p *= beta;
    p += r;

    if (r_sq < r_sq_stop_eps){
      break;
    }
  }

  const torch::Tensor lse = mul_A_T(x, y, xi, x_width, chunk_size);

  return lse;
}
